src/ccache/util/LockFile.cpp

```diff
@@ -236,16 +238,31 @@ LockFile::do_acquire(const bool blocking)
     const auto now = TimePoint::now();
     const auto my_content =
       FMT("{}-{}.{}", content_prefix, now.sec(), now.nsec_decimal_part());
-
+#  ifdef __CYGWIN__
+    // Cygwin-specific file-based lock
+    int fd = open(m_lock_file.c_str(), O_WRONLY | O_CREAT | O_EXCL, 0666);
+    if (fd != -1) {
+      // Lock file successfully created, write the content and close the file
+      write(fd, my_content.c_str(), my_content.size());
+      close(fd); // Lock acquired
+      return true;
+    }
+#  else
     if (fs::create_symlink(my_content, m_lock_file)) {
       // We got the lock.
       return true;
     }
+#  endif
 
     int saved_errno = errno;
     if (saved_errno == ENOENT) {
       // Directory doesn't exist?
+#  ifdef __CYGWIN__
+      if (!fs::exists(m_lock_file.parent_path())
```

Comment:
For my own understanding: Can `open(..., ... O_CREAT ...)` set `errno` to `ENOENT` in msys2 for other reasons than _"a directory component in pathname does not exist or is a dangling symbolic link"_? If not, I guess the `fs::exists` call is superfluous?

Reply:
The fs::exists was added because it solves the infinite loop problem, apparently in some cases, it returns `true` even when the path exists.